### ISSUE TITLE
De-couple FIFE from Guichan

### ISSUE THREAD

**Issue by prock-fife** [fifengine/fifengine#549]
This task is to de-couple FIFE from Guichan and to add an interface to allow the easy adoption of other GUI libraries (such as CEGUI).  In the process it will make us true to our name (F for Flexible :D).

**Comment by jakoch** [comment]
❓ Does de-coupling include moving the PyChan layer (https://github.com/fifengine/fifengine/tree/master/engine/python/fife/extensions/pychan) from the fifengine repo to the fifechan repo?

If this isn't on the map, yet, i would suggest adding it:
- because a change to the fifechan cpp code will probably always need a tweak to the generated interface and that means the commits and PRs land in the same repo 
   - currently, the cpp changes go to fifechan and changes to the wrapper code go to fifechan
- the build artifacts are "libfifechan" and "pyfifechan" (short PyChan) and they could be build from one repository

**Comment by prock-fife** [comment]
I agree that we need to de-couple it.  There are some challenges however.   Pychan itself depends on FIFE itself so the major work might be removing that dependency.

**Comment by LinuxDonald** [comment]
@prock should I than stopp merging the pychan branch?

**Comment by Beliaar** [comment]
Decoupling would certainly be easier if we would split off fife into a shared fife library, a python extension module and the gui managers as seperate python modules that link to fife. This way fifechan could still use fife classes (optionally, of course, for those who don't want/need fife).

**Comment by prock-fife** [comment]
@LinuxDonald it's really 2 separate issues.  De-coupling "pyfifechan" from FIFE would enable, as @Beliaar mentioned the ability for clients to use  "pyfifechan" without needing to use FIFE.  What @helios2000 was trying to do was move the functionality from "pyfifechan" to fifechan itself.  The reason he was doing this was because pychan itself was becoming hard to maintain and generally it's fairly buggy.  Because we maintain fifechan it make more sense to put functionality into it instead of the python module itself.  This also gives better support to people who don't want to use python and would prefer using the C++ libraries.

Ideally we would do both... de-couple pychan from FIFE AND move the extra functionality from pychan to fifechan.  The question is in what order makes the most sense.  @helios2000 had decided to work on moving functionality from pychan to fifechan first.

**Comment by jakoch** [comment]
That's a good explanation of the situation and where we are heading in this area.
I suggest we proceed doing the things in the order (move functionality first, then split out). 
Changing the order would mean to re-target existing pull requests to other repos, 
which is a lot of additional maintenance work.

- Remove the hard dependency on the Guichan GUI library from FIFE's core 
- Introduce an abstraction layer (interface) that allows swapping in alternative GUI libraries (e.g., CEGUI)

**Comment by prock-fife** [comment]
> I suggest we proceed doing the things in the order (move functionality first, then split out). 
Agreed.  Let's proceed with #843 and park this issue until that is completed or at least the functionality that @helios2000 has migrated has been fully integrated into master.

**Comment by jakoch** [comment]
As this is quite old, i let ChatGPT provide a summary:

#### Issue Summary

**Proposal by prock-fife**: Create a GUI abstraction interface to allow plugging in different GUI libraries (CEGUI, etc.) and move all Guichan-specific code behind it, making the "F" in FIFE (Flexible) true.

**Implementation**:
1. All Fifechan code was physically isolated to engine/core/gui/guichan/
2. Engine was modified to operate without a GUI manager entirely
3. Python bindings split into two modules: 
   - fife.py (core)
   - fifechan.py (GUI) fifechan module optional at build time via --without-fifechan
4. PychanBasicApplication introduced as a fifechan-aware base class
   - BasicApplication became GUI-agnostic
5. Font requirements removed from renderer constructors (CoordinateRenderer, FloatingTextRenderer)
   - fonts passed by clients
6. Python-side decoupling: 
  - fife_settings decoupled from default settings GUI panel using inheritance

This was a straightforward architectural refactoring with the core design decision being "isolate behind an interface + make optional at build time" rather than fully abstracting for hot-swappable GUI backends.

----

The idea sounds appealing at first glance, but I think the practical angle is worth considering:
- Even if we pulled in CEGUI, ImGUI, or Raylib, we'd still need to write SWIG bindings for them in this repo. none of them ship SWIG bindings upstream.
  - The abstraction argument falls apart when every alternative library still needs SWIG bindings maintained in this repo anyway.
- That raises the question: why should we provide a Python layer for fifechan in the fifechan repo, or add SWIG there?
- Moving pychan or adding swig into the fifechan repo doesn't seem like the right approach
- I think the extra functionality pychan provides should move into fifechan's C++ side instead, keeping pychan as just a thin Python layer on top. This also avoids fragmentation of the binding logic across repos.. which is hard enough to maintain in one repo.

So this thread really only yields one concrete task (except someone else would like to work on the original approach and move everything into fifechan): 
- [ ] migrate pychan's extra functionality into fifechan itself, leaving pychan as a thin wrapper
